### PR TITLE
Convert `newtype_index` to a proc macro

### PR DESCRIPTION
**Pull request by Aaron1011**
The `macro_rules!` implementation was becomng excessively complicated,
and difficult to modify. The new proc macro implementation should make
it much easier to add new features (e.g. skipping certain `#[derive]`s)

### CODE REVIEW

File: compiler/rustc_macros/src/lib.rs [hunk 1]
```diff
@@ -24,6 +26,27 @@ pub fn symbols(input: TokenStream) -> TokenStream {
...
 }
 
+/// Creates a struct type `S` that can be used as an index with
+/// `IndexVec` and so on.
+///
+/// There are two ways of interacting with these indices:
+///
+/// - The `From` impls are the preferred way. So you can do
+///   `S::from(v)` with a `usize` or `u32`. And you can convert back
+///   to an integer with `u32::from(s)`.
```
Comment (nnethercote) [inline compiler/rustc_macros/src/lib.rs:36]: I know this comment is pre-existing, but while we're here: is this advice correct? I personally prefer `S::new(v)` and `s.index()`, and I feel like I've seen those forms more often.

File: compiler/rustc_macros/src/newtype.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+            impl From<u32> for #name {
+                #[inline]
+                fn from(value: u32) -> Self {
+                    Self::from_u32(value)
+                }
+            }
+
+            #encodable_impls
+            #debug_impl
+
```
Comment (cjgillot) [inline compiler/rustc_macros/src/newtype.rs:316]: Stray newline?

**Review by cjgillot — COMMENTED** [review]
This is a great simplification to the expansion!

File: compiler/rustc_macros/src/lib.rs (hunk 1 above)
Comment (Aaron1011) [reply compiler/rustc_macros/src/lib.rs:36]: I'm not sure - I think this could be dealt with in a follow-up issue, since it's just a style reccomendation.

**Comment by Aaron1011** [comment]
@cjgillot I've addressed your comments

**Comment by matthiaskrgr** [comment]
might be reponsible for these ci failures: https://github.com/rust-lang/rust/pull/94167#issuecomment-1046123420 
@bors rollup=never

**Comment by Mark-Simulacrum** [comment]
Likely just noise -- externs is currently a little noisy (and we haven't absorbed that noisiness into our statistical estimation yet). Seems to be due to https://github.com/rust-lang/rust/pull/93839, which makes me suspect this is related to PGO or inlining decisions of some kind. https://github.com/rust-lang/rust/pull/94373 might help.

**Pull request merged**